### ISSUE TITLE
mp4 file get downloaded even if i have a non fully downloaded m4a file

### ISSUE THREAD

**Issue by Sopor** [spaam/svtplay-dl#1360]
If i cancel a download when it is downloading the m4a file and i then run the same command again it will jump to the mp4 file. This behaviour worked when svtplay-dl downloaded one file that contained both audio and video, but now it will only leave a broken audio file.

```
svtplay-dl.exe https://www.svtplay.se/video/30570370/snoanglar/snoanglar-sasong-1-avsnitt-1
INFO: Outfile: snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.srt
INFO: Selected to download dash, bitrate: 6043 format: h264-51
INFO: Outfile: snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.m4a
[086/894][=======.........................................................................] ETA: 0:00:21

CTRL+C

svtplay-dl.exe https://www.svtplay.se/video/30570370/snoanglar/snoanglar-sasong-1-avsnitt-1
INFO: Outfile: snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.srt
WARNING: File (snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.srt) already exists. Use --force to overwrite
INFO: Selected to download dash, bitrate: 6043 format: h264-51
INFO: Outfile: snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.m4a
WARNING: File (snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.m4a) already exists. Use --force to overwrite
INFO: Outfile: snoanglar.s01e01.avsnitt.1-133ba4b-svtplay.mp4
[024/894][==..............................................................................] ETA: 0:01:42
```

I know it tells me to use `--force` but it won't tell me if the audio file is completely downloaded or not. Audio and video should be treated as one and if the m4a file exists it shouldn't continue with the mp4 file. I then will get a corrupted merged file if let it finished.

I don't know if there are any switches to tell svtplay-dl to only download the audio/video? Maybe that could be something to add if i already have the audio i could tell svtplay-dl to only download the video?

**Comment by pythonuser3856** [comment]
In this area of functionality I make active use of that download will move to download the .mp4-part when it detects presence of an .m4a-file.
It gives me the opportunity to perform download in parallel of the .m4a and .mp4, thus shorten the time of a full download given that there is sufficient bandwidth to match.
For that reason I'd very much like to keep this functionality of download in parallel when the source is an .m4a- and an .mp4-file.

One way that this could be addressed is to let svtplay-dl inherently perform parallel download in the scenario. If the download is interrupted the svtplay-dl would have the option to dispose of partially download files. One way to represent a state of partially downloaded files is to use temporary file names until download completion.

**Comment by bohelax** [comment]
A reason not to do parallel downloads of .m4a and .mp4 is that apparently the servers sometimes have to create the dash files from some sources, and if they are busy this can cause breaks (gaps) in the transmission (as seen by a network monitoring tool and by sometimes an excessive download time). I have found that when this happens the result is often unreliable. I then discard that download and simply run it again - then it seems that the servers have cashed the 'to dash' conversion.

Btw if/when I interrupt a download I always discard everything before trying again.
//BoH

**Comment by spaam** [comment]
its a known bug for years and it does not break any normal downloads. 

1) adding `part` prefix to the filename like the end of it (eg julmust_vs_påskmust-the.same_or_not.mp4.part). when we see the `part` in the filename next time we download the same file, should we just remove the file and create a new one? 🤔 

2) handle the ctrl+c and remove the file that we are downloading at that moment?  🤔

**Comment by pythonuser3856** [comment]
Given current logic of svtplay-dl I agree with that it's unreliable to perform parallel download; I've noticed that downloading soon after release of a weblink at SVTplay is slower than otherwise and my guess is that the video at the point in time has not yet been cached by the Akamai servers.

The unreliable outcome from using svtplay-dl to perform parallel download today comes from that this extended use of svtplay-dl depends on timing.
A different and more reliable approach is to base it on states to control parallel download of .m4a and .mp4, and use an execution rendevouz to wait for completion of both before continuing.

If either of the downloads fails it's an abnormal condition. As long as svtplay-dl at that  point in time is in control of execution, it would be feasible for svtplay-dl to perform a cleanup of any temporary files before exiting. A cleanup while still executing enables svtplay-dl to use its internal information to be aware of what to remove (such as a use of a random-ID as part of a temporary file name).

If svtplay-dl instead crashes it could at next start rely on state information stored in the file system to become aware of what to remove when it's started next time. Using the substring "part" whithin tha file name is one solution to achieve that.
Or just leave it to the user to perform the file system cleanup.

I'm aware of that this reasoning is about adding logic to make the use reliable and supported and I'm fine with the functionality of svtplay-dl as it is today.
However, I'd like to see is that my extended use to perform parallel download of .m4a and .mp4 remains as a possible option for me.

**Comment by bohelax** [comment]
Another possible way to make svtplay-dl more 'fail safe' could be to make '--force' the default, and maybe introduce '--noforce' for those who know what they are doing.
//BoH

**Comment by pythonuser3856** [comment]
I've tested the snaphot-version 20210509_195246.
Two comments:
1) it prevents me from doing parallel download of .m4a,
2) it crashes.

Trace at testing using the link https://www.svtplay.se/video/29055502/rio-rocket-en-osannolik-speedwayresa
( using snapshot-version 20210422_175908 is successful )

py c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\__main__.py --subtitle https://www.svtplay.se/video/29055502/rio-rocket-en-osannolik-speedwayresa
INFO: Outfile: rio.rocket-en.osannolik.speedwayresa.rio.rocket-en.osannolik.speedwayresa-32c6737-svtplay.srt
INFO: Outfile: rio.rocket-en.osannolik.speedwayresa.rio.rocket-en.osannolik.speedwayresa-32c6737-svtplay.mp4
INFO: Selected to download dash, bitrate: 6032 format: h264
[252/252][========================================] ETA: 0:00:00
[252/252][========================================] ETA: 0:00:00
Traceback (most recent call last):
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\__main__.py", line 15,in <module>
    svtplay_dl.main()
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\__init__.py", line 69, in main
    get_media(urls[0], config, __version__)
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\getmedia.py", line 62, in get_media
    get_one_media(stream)
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\getmedia.py", line 238, in get_one_media
    post.merge()
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\postprocess\__init__.py", line 120, in merge
    _, stdout, stderr = run_program(cmd, False)  # return 1 is good here.
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\proc.py", line 6, in run_program
    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
  File "C:\Program Files\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Python37\lib\subprocess.py", line 1119, in _execute_child
    args = list2cmdline(args)
  File "C:\Program Files\Python37\lib\subprocess.py", line 530, in list2cmdline
    needquote = (" " in arg) or ("\t" in arg) or not arg
TypeError: argument of type 'WindowsPath' is not iterable

**Comment by spaam** [comment]
just curious , is that the .exe or the svtplay-dl thing? feels like stuff missing from the path  `svtplay-dl_sshot_20210509_195246py_main_.py`  🤔 

interesting crash.. it works great on macos and linux

**Comment by spaam** [comment]
it works on python on windows for me with python3.8.  what version do you have? 3.7? 3.6?

**Comment by pythonuser3856** [comment]
I'm using the python code variant and running it on python 3.7.4

What I've used with previous versions of svtplay-dl is to start download of the .m4a-file with one instance of svtplay-dl.
In parallel I use a different instance of svtplay-dl in the same directory. The second invocation of svtplay-dl find the locked .m4a-file being processed by the first svtplay-dl instance and proceed to download the .mp4-file
Download of the .m4a-file concludes in advance of the .mp4-file and the second svtplay-dl instance use it as input to create a result.

**Comment by pythonuser3856** [comment]
I re-fetched and re-installed snapshot 202105-09_195246
I also fetched snapshot version 20210509_001313
I get the same crash as before with both versions

**Comment by pythonuser3856** [comment]
Upgraded installed python to version 3.8.10 and running the python code version of svtplay-dl.
Still a crash when executing either of snap-shot version 2021-0509_195246  or 20210509_001313:

-------------------
2021-05-10 13:10:33
-------------------
py c:\svtplaydll\svtplay-dl_sshot_20210509_19524
6py\__main__.py --subtitle https://www.svtplay.se/video/29055502/rio-rocket-en-o
sannolik-speedwayresa
INFO: Outfile: rio.rocket-en.osannolik.speedwayresa.rio.rocket-en.osannolik.spee
dwayresa-32c6737-svtplay.srt
WARNING: File (rio.rocket-en.osannolik.speedwayresa.rio.rocket-en.osannolik.spee
dwayresa-32c6737-svtplay.srt) already exists. Use --force-subtitle to overwrite
INFO: Outfile: rio.rocket-en.osannolik.speedwayresa.rio.rocket-en.osannolik.spee
dwayresa-32c6737-svtplay.mp4
INFO: Selected to download dash, bitrate: 6032 format: h264
[252/252][========================================] ETA: 0:00:00
[252/252][========================================] ETA: 0:00:00
Traceback (most recent call last):
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\__main__.py", line 15,in <module>
    svtplay_dl.main()
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\__init__.py", line 69, in main
    get_media(urls[0], config, __version__)
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\getmedia.py", line 62, in get_media
    get_one_media(stream)
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\getmedia.py", line 238, in get_one_media
    post.merge()
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\postprocess\__init__.py", line 120, in merge
    _, stdout, stderr = run_program(cmd, False)  # return 1 is good here.
  File "c:\svtplaydll\svtplay-dl_sshot_20210509_195246py\svtplay_dl\utils\proc.py", line 6, in run_program
    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, stdin=subprocess.PIPE)
  File "C:\Program Files\Python37\lib\subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "C:\Program Files\Python37\lib\subprocess.py", line 1119, in _execute_child
    args = list2cmdline(args)
  File "C:\Program Files\Python37\lib\subprocess.py", line 530, in list2cmdline
    needquote = (" " in arg) or ("\t" in arg) or not arg
TypeError: argument of type 'WindowsPath' is not iterable

**Comment by pythonuser3856** [comment]
I must have had some half-baked installation after upgrade of python to 3.8.10.
The latest svtplay-dl now works without crashing.
But the solution prevents performing parallel download of .mp4 and .m4a using two instances of svtplay-dl running simultajneously towards the same directory, thus prevents to use that tweak to speed up download.

**Comment by pythonuser3856** [comment]
Good proposal but it does not work when I tried.
- Download of audio concludes leaving a .temp-file (since it's blocked by the parallel video download)
- Download of video concludes with a message of merged audio but the .mp4-file is video only.
example with use of "--only-video" with a .temp-file from "--only-audio --force" present:
INFO: Outfile: curling.vm.final-bd1c23d-svtplay.mp4
INFO: Selected to download dash, bitrate: 3256 format: h264
[1492/1492][========================================] ETA: 0:00:00
INFO: Merge audio and video into curling.vm.final-bd1c23d-svtplay.mp4
INFO: Merging done, removing old files.

I guess I can live with longer download times, it was just a non-supported tweak I used.

**Comment by spaam** [comment]
i added this feature to stop downloading files that we already have parts of or fully downloaded in `3.8` if there is still an issue or so, open a new issue.
